### PR TITLE
(TWILL-153) Honor actual resource size of the container

### PR DESCRIPTION
**Pull request by chtyim**
- Determine the -Xmx based on the actual size of the container
  - it can be smaller or bigger than the one requested in the TwillSpec
- Refactor resource specification for AM
  - A forward looking change for TWILL-90
- Simple code cleanup to get rid of code warning from IDE.

### CODE REVIEW

File: twill-zookeeper/src/test/java/org/apache/twill/zookeeper/ZKClientTest.java [hunk 1]
```diff
@@ -188,12 +189,21 @@ public void process(WatchedEvent event) {
       client.startAndWait();
 
       try {
-        final BlockingQueue<Watcher.Event.EventType> events = new LinkedBlockingQueue<Watcher.Event.EventType>();
+        final BlockingQueue<Watcher.Event.EventType> events = new LinkedBlockingQueue<>();
```
Comment (hsaputra) [inline twill-zookeeper/src/test/java/org/apache/twill/zookeeper/ZKClientTest.java]: Is the update of this test related to the respecting container resource?

File: twill-yarn/src/main/hadoop20/org/apache/twill/internal/yarn/Hadoop20YarnAppClient.java [hunk 2]
```diff
@@ -70,8 +72,8 @@ public Hadoop20YarnAppClient(Configuration configuration) {
   }
 
   @Override
-  public ProcessLauncher<ApplicationId> createLauncher(TwillSpecification twillSpec,
-                                                       @Nullable String schedulerQueue) throws Exception {
+  public ProcessLauncher<ApplicationMasterInfo> createLauncher(TwillSpecification twillSpec,
+                                                               @Nullable String schedulerQueue) throws Exception {
```
Comment (hsaputra) [inline twill-yarn/src/main/hadoop20/org/apache/twill/internal/yarn/Hadoop20YarnAppClient.java]: Small nit on formatting, is that extra tab chars? Wondering why it has different green shade?

**Comment by hsaputra** [comment]
Did a pass and add small comments.

The rest LGTM +1

File: twill-yarn/src/main/hadoop20/org/apache/twill/internal/yarn/Hadoop20YarnAppClient.java (hunk 2 above)
Comment (chtyim) [inline twill-yarn/src/main/hadoop20/org/apache/twill/internal/yarn/Hadoop20YarnAppClient.java]: Just the change in alignment. Should be all spaces, no tab

File: twill-zookeeper/src/test/java/org/apache/twill/zookeeper/ZKClientTest.java (hunk 1 above)
Comment (chtyim) [inline twill-zookeeper/src/test/java/org/apache/twill/zookeeper/ZKClientTest.java]: This line is just code cleanup. The modification to this test case is a fix to make the race condition go away. This test used to fail easily in travis

**Comment by hsaputra** [comment]
Will merge this later today if no more comments.

File: twill-core/src/main/java/org/apache/twill/internal/utils/Resources.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  /**
+   * Computes the max heap size for a JVM process.
+   *
+   * @param containerMemory memory in MB of the container memory memory.
+   *                        It is the maximum memory size allowed for the process.
+   * @param nonHeapMemory memory in MB that needs to be reserved for non JVM heap memory for the process.
+   * @param minHeapRatio minimum ratio for heap to non-heap memory.
+   * @return memory in MB representing the max heap size for a JVM process.
+   */
+  public static int computeMaxHeap(int containerMemory, int nonHeapMemory, double minHeapRatio) {
```
Comment (gsps1) [inline twill-core/src/main/java/org/apache/twill/internal/utils/Resources.java]: ``computeMaxHeapSize`` is a better name ?

File: twill-core/src/main/java/org/apache/twill/internal/utils/Resources.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Utility class to help adjusting container resource requirement.
+ */
+public final class Resources {
+
+  /**
+   * Computes the max heap size for a JVM process.
+   *
+   * @param containerMemory memory in MB of the container memory memory.
```
Comment (poornachandra) [inline twill-core/src/main/java/org/apache/twill/internal/utils/Resources.java]: `... container memory memory` -  memory repeated.

Comment (chtyim) [inline twill-core/src/main/java/org/apache/twill/internal/utils/Resources.java]: Removed.

File: twill-core/src/main/java/org/apache/twill/internal/utils/Resources.java (hunk 3 above)
Comment (chtyim) [inline twill-core/src/main/java/org/apache/twill/internal/utils/Resources.java]: Renamed.

**Comment by hsaputra** [comment]
Ah ok looks like @chtyim has merged it.

**Pull request closed without merging**